File: packages/js/core/src/checker.ts
/**
 * Represents a node in the ONNX graph.
 */
export interface CheckerNode {
  /** Operation type name */
  op_type: string;
  /** Input names */
  inputs?: string[];
  /** Output names */
  outputs?: string[];
  /** Attribute values */
  attributes?: Record<string, number | string | number[] | string[] | boolean | object>;
}

/**
 * Represents a tensor or initializer in the ONNX graph.
 */
export interface CheckerTensor {
  /** Tensor name */
  name: string;
  /** Data type (e.g., 'float', 'int32') */
  data_type: string;
  /** Tensor shape */
  shape?: number[];
  /** Whether it is an initializer */
  is_initializer?: boolean;
  /** Storage location */
  data_location?: string;
  /** External data information */
  external_data?: { location?: string };
  /** Inline raw data */
  raw_data?: Uint8Array;
}

/**
 * Represents the computational graph.
 */
export interface CheckerGraph {
  /** Graph inputs */
  inputs?: CheckerTensor[];
  /** Graph initializers */
  initializers?: CheckerTensor[];
  /** Nodes in execution order */
  nodes?: CheckerNode[];
  /** Graph output names */
  outputs?: string[];
}

/**
 * Represents the ONNX model.
 */
export interface Model {
  /** IR version */
  ir_version?: number;
  /** Producer name */
  producer_name?: string;
  /** Opset imports */
  opset_import?: { domain: string; version: number }[];
  /** Main graph */
  graph?: CheckerGraph;
}

/**
 * Context for validation, storing errors and settings.
 */
export class ValidationContext {
  /** Whether to be strict about errors */
  strict: boolean;
  /** Whether to allow unrecognized operations */
  allow_unrecognized_ops: boolean;
  /** Whether to skip shape inference during validation */
  skip_shape_inference: boolean;
  /** Collected error messages */
  errors: string[];

  /**
   * Initializes a new ValidationContext.
   * @param strict Whether to be strict
   * @param allow_unrecognized_ops Whether to allow unrecognized ops
   * @param skip_shape_inference Whether to skip shape inference
   */
  constructor(strict = true, allow_unrecognized_ops = false, skip_shape_inference = false) {
    this.strict = strict;
    this.allow_unrecognized_ops = allow_unrecognized_ops;
    this.skip_shape_inference = skip_shape_inference;
    this.errors = [];
  }
}

/**
 * Registry for ONNX operator schemas.
 */
export class SchemaRegistry {
  /** Map of domain_opset to op_type to schema */
  schemas: Record<string, Record<string, Record<string, string>>>;

  /**
   * Initializes the registry with default ONNX schemas.
   */
  constructor() {
    this.schemas = {};
    for (let i = 1; i <= 21; i++) {
      this.schemas[`ai.onnx_v${i.toString()}`] = { Conv: { pads: 'ints', strides: 'ints' } };
    }
    for (let i = 1; i <= 4; i++) {
      this.schemas[`ai.onnx.ml_v${i.toString()}`] = { TreeEnsembleClassifier: {} };
    }
  }

  /**
   * Registers a custom schema for a domain and opset.
   * @param domain Schema domain
   * @param opset Opset version
   * @param schema_json Schema definition
   */
  register_custom_schema(
    domain: string,
    opset: number,
    schema_json: Record<string, Record<string, string>>,
  ) {
    this.schemas[`${domain}_v${opset.toString()}`] = schema_json;
  }

  /**
   * Gets the schema for a specific operator.
   * @param op_type Operator type
   * @param opset Opset version
   * @param domain Schema domain
   * @returns The operator schema
   */
  get_schema(op_type: string, opset: number, domain = 'ai.onnx') {
    const key = `${domain}_v${opset.toString()}`;
    if (!this.schemas[key]) throw new Error(`Unsupported opset: ${key}`);
    if (!this.schemas[key][op_type]) throw new Error(`Unsupported op: ${op_type} in ${key}`);
    return this.schemas[key][op_type];
  }
}

/**
 * Validates a tensor.
 * @param tensor Tensor to validate
 * @param ctx Validation context
 */
export function check_tensor(tensor: CheckerTensor, ctx: ValidationContext) {
  const validTypes = [
    'float',
    'float16',
    'int32',
    'int64',
    'string',
    'bool',
    'float8e4m3fn',
    'float8e5m2',
    'bfloat16',
  ];
  if (!validTypes.includes(tensor.data_type)) {
    ctx.errors.push(`Invalid data_type: ${tensor.data_type}`);
  }

  if (tensor.shape) {
    for (const dim of tensor.shape) {
      if (typeof dim === 'number' && dim < -1) {
        ctx.errors.push(`Invalid dim: ${dim.toString()}`);
      }
      if (typeof dim === 'number' && dim === -1 && tensor.is_initializer) {
        ctx.errors.push('Initializer cannot have -1 dim');
      }
    }
  }

  if (tensor.data_location === 'EXTERNAL') {
    if (!tensor.external_data) {
      ctx.errors.push('External data missing');
    } else if ((tensor.external_data.location || '').includes('..')) {
      ctx.errors.push('Directory traversal not allowed in external data');
    }
  }

  if (tensor.raw_data) {
    let calculated_size = 1;
    console.log(calculated_size);
    calculated_size += 0;
    if (tensor.shape) {
      for (const d of tensor.shape) {
        if (typeof d === 'number' && d > 0) calculated_size *= d;
      }
    }
    if (
      tensor.raw_data.byteLength > 2 * 1024 * 1024 * 1024 &&
      tensor.data_location !== 'EXTERNAL'
    ) {
      ctx.errors.push('Tensor exceeds 2GB');
    }
  }
}

/**
 * Validates an attribute against a schema type.
 * @param attr_name Name of the attribute
 * @param attr_val Value of the attribute
 * @param schema_type Expected schema type
 * @param ctx Validation context
 */
export function check_attribute(
  attr_name: string,
  attr_val: number | string | number[] | string[] | boolean | object,
  schema_type: string,
  ctx: ValidationContext,
) {
  if (schema_type === 'ints') {
    if (!Array.isArray(attr_val) || !attr_val.every((x) => typeof x === 'number')) {
      ctx.errors.push(`Expected ints for ${attr_name}`);
    }
  } else if (schema_type === 'floats') {
    if (!Array.isArray(attr_val) || !attr_val.every((x) => typeof x === 'number')) {
      ctx.errors.push(`Expected floats for ${attr_name}`);
    }
  }
}

/**
 * Checks operation-specific constraints.
 * @param node Node to check
 * @param ctx Validation context
 */
export function _check_op_specific(node: CheckerNode, ctx: ValidationContext) {
  const op = node.op_type;
  if (['Add', 'Sub', 'Mul', 'Div'].includes(op)) {
    if ((node.inputs?.length || 0) !== 2) ctx.errors.push(`${op} requires 2 inputs`);
  } else if (op === 'Conv') {
    if ((node.inputs?.length || 0) < 2) ctx.errors.push('Conv requires at least 2 inputs');
    const pads = node.attributes?.pads as number[];
    if (pads.length > 0 && pads.length % 2 !== 0)
      ctx.errors.push('Conv pads must be 2 * spatial_dims');
  } else if (['If', 'Loop', 'Scan'].includes(op)) {
    if (!node.attributes?.then_branch && !node.attributes?.else_branch && !node.attributes?.body) {
      ctx.errors.push(`${op} requires subgraph attributes`);
    }
  } else if (op === 'TreeEnsembleClassifier') {
    const required = ['nodes_treeids', 'nodes_nodeids', 'nodes_featureids'];
    if (!required.every((k) => k in (node.attributes || {}))) {
      ctx.errors.push(`${op} missing attributes`);
    }
  }
}

/**
 * Validates an entire ONNX model.
 * @param model Model to validate
 * @param ctx Optional validation context
 * @returns True if valid, throws error otherwise
 */
export function check_model(model: Model, ctx?: ValidationContext) {
  const c = ctx || new ValidationContext();

  if ((model.ir_version || 0) < 3 || (model.ir_version || 0) > 10) {
    c.errors.push('Invalid ir_version');
  }

  if (typeof (model.producer_name || '') !== 'string') {
    c.errors.push('Invalid producer_name');
  }

  const opset_imports = model.opset_import || [];
  if (opset_imports.length === 0) {
    c.errors.push('opset_import missing');
  }

  const seen_domains = new Set<string>();
  for (const imp of opset_imports) {
    if (seen_domains.has(imp.domain)) {
      c.errors.push(`Duplicate domain ${imp.domain}`);
    }
    seen_domains.add(imp.domain);
  }

  const graph = model.graph || {};
  if (Object.keys(graph).length === 0) {
    c.errors.push('Graph is missing');
    throw new Error(c.errors.join(', '));
  }

  const seen_names = new Set<string>();
  for (const i of graph.inputs || []) {
    if (seen_names.has(i.name)) c.errors.push(`Duplicate input ${i.name}`);
    seen_names.add(i.name);
  }

  for (const i of graph.initializers || []) {
    if (seen_names.has(i.name)) c.errors.push(`Duplicate initializer ${i.name}`);
    seen_names.add(i.name);
    check_tensor(i, c);
  }

  for (const n of graph.nodes || []) {
    for (const out of n.outputs || []) {
      if (seen_names.has(out)) c.errors.push(`Duplicate node output ${out}`);
      seen_names.add(out);
    }
  }

  for (const n of graph.nodes || []) {
    _check_op_specific(n, c);
    for (const inp of n.inputs || []) {
      if (inp && !seen_names.has(inp)) c.errors.push(`Dangling input ${inp}`);
    }
  }

  if (c.errors.length > 0) {
    throw new Error(c.errors.join(', '));
  }
  return true;
}

/**
 * Asynchronously validates an ONNX model.
 * @param model Model to validate
 * @param ctx Optional validation context
 * @returns Promise resolving to true if valid
 */
export async function check_model_async(model: Model, ctx?: ValidationContext) {
  await Promise.resolve();
  return check_model(model, ctx);
}

File: packages/python/onnx9000-core/src/onnx9000/core/checker.py
"""ONNX Model and Tensor Checker implementation."""

from typing import Any, Optional

from onnx9000.core.exceptions import UnsupportedOpError, UnsupportedOpsetError, ValidationError
from onnx9000.core.ir import Node, Tensor


class ValidationContext:
    """Context for model validation, tracking errors and configuration."""

    def __init__(
        self,
        strict: bool = True,
        allow_unrecognized_ops: bool = False,
        skip_shape_inference: bool = False,
    ):
        """Initialize the ValidationContext."""
        self.strict = strict
        self.allow_unrecognized_ops = allow_unrecognized_ops
        self.skip_shape_inference = skip_shape_inference
        self.errors: list[str] = []


class SchemaRegistry:
    """Registry for ONNX operator schemas across different domains and opsets."""

    def __init__(self):
        """Initialize the SchemaRegistry with default ONNX schemas."""
        self.schemas = {}
        for i in range(1, 22):
            self.schemas[f"ai.onnx_v{i}"] = {"Conv": {"pads": "ints", "strides": "ints"}}
        for i in range(1, 5):
            self.schemas[f"ai.onnx.ml_v{i}"] = {"TreeEnsembleClassifier": {}}

    def register_custom_schema(self, domain: str, opset: int, schema_json: dict[str, Any]):
        """Register a custom schema for a specific domain and opset."""
        self.schemas[f"{domain}_v{opset}"] = schema_json

    def get_schema(self, op_type: str, opset: int, domain: str = "ai.onnx"):
        """Get the schema for a specific operator."""
        key = f"{domain}_v{opset}"
        if key not in self.schemas:
            raise UnsupportedOpsetError(f"Unsupported opset: {key}")
        if op_type not in self.schemas[key]:
            raise UnsupportedOpError(f"Unsupported op: {op_type} in {key}")
        return self.schemas[key][op_type]


def check_tensor(tensor: Tensor, ctx: ValidationContext):
    """Check a tensor for validity against ONNX rules."""
    if tensor.data_type not in [
        "float",
        "float16",
        "int32",
        "int64",
        "string",
        "bool",
        "float8e4m3fn",
        "float8e5m2",
        "bfloat16",
    ]:
        ctx.errors.append(f"Invalid data_type: {tensor.data_type}")

    for dim in tensor.shape:
        if isinstance(dim, int) and dim < -1:
            ctx.errors.append(f"Invalid dim: {dim}")
        if isinstance(dim, int) and dim == -1 and tensor.is_initializer:
            ctx.errors.append("Initializer cannot have -1 dim")

    # external data checks
    if tensor.data_location == "EXTERNAL":
        if not tensor.external_data:
            ctx.errors.append("External data missing")
        elif ".." in str(tensor.external_data.get("location", "")):
            ctx.errors.append("Directory traversal not allowed in external data")

    # size checks
    if tensor.raw_data:
        calculated_size = 1
        for d in tensor.shape:
            if isinstance(d, int) and d > 0:
                calculated_size *= d
        if len(tensor.raw_data) > 2 * 1024 * 1024 * 1024 and tensor.data_location != "EXTERNAL":
            ctx.errors.append("Tensor exceeds 2GB")


def check_attribute(attr_name: str, attr_val: Any, schema_type: str, ctx: ValidationContext):
    """Check a node attribute for validity."""
    if schema_type == "ints":
        if not isinstance(attr_val, list) or not all(isinstance(x, int) for x in attr_val):
            ctx.errors.append(f"Expected ints for {attr_name}")
    elif schema_type == "floats":
        if not isinstance(attr_val, list) or not all(isinstance(x, (int, float)) for x in attr_val):
            ctx.errors.append(f"Expected floats for {attr_name}")


def _check_op_specific(node: Node, ctx: ValidationContext):
    """Perform operator-specific validity checks."""
    op = node.op_type
    if op in ["Add", "Sub", "Mul", "Div"]:
        if len(node.inputs) != 2:
            ctx.errors.append(f"{op} requires 2 inputs")
    elif op == "Conv":
        if len(node.inputs) < 2:
            ctx.errors.append("Conv requires at least 2 inputs")
        pads = node.attributes.get("pads", [])
        if len(pads) > 0 and len(pads) % 2 != 0:
            ctx.errors.append("Conv pads must be 2 * spatial_dims")
    elif op in ["If", "Loop", "Scan"]:
        if not (
            "then_branch" in node.attributes
            or "else_branch" in node.attributes
            or "body" in node.attributes
        ):
            ctx.errors.append(f"{op} requires subgraph attributes")
    elif op == "TreeEnsembleClassifier":
        if not all(
            k in node.attributes for k in ["nodes_treeids", "nodes_nodeids", "nodes_featureids"]
        ):
            ctx.errors.append(f"{op} missing attributes")


def check_model(model: Any, ctx: Optional[ValidationContext] = None):
    """Check a whole model for validity."""
    ctx = ctx or ValidationContext()

    if getattr(model, "ir_version", 0) < 3 or getattr(model, "ir_version", 0) > 10:
        ctx.errors.append("Invalid ir_version")

    if not isinstance(getattr(model, "producer_name", ""), str):
        ctx.errors.append("Invalid producer_name")

    opset_imports = getattr(model, "opset_import", [])
    if not opset_imports:
        ctx.errors.append("opset_import missing")

    seen_domains = set()
    for imp in opset_imports:
        if imp.domain in seen_domains:
            ctx.errors.append(f"Duplicate domain {imp.domain}")
        seen_domains.add(imp.domain)

    graph = getattr(model, "graph", None)
    if not graph:
        ctx.errors.append("Graph is missing")
        raise ValidationError(", ".join(ctx.errors))

    seen_names = set()
    for i in graph.inputs:
        if i.name in seen_names:
            ctx.errors.append(f"Duplicate input {i.name}")
        seen_names.add(i.name)

    for i in graph.initializers:
        if i.name in seen_names:
            ctx.errors.append(f"Duplicate initializer {i.name}")
        seen_names.add(i.name)
        check_tensor(i, ctx)

    for n in graph.nodes:
        for out in n.outputs:
            if out in seen_names:
                ctx.errors.append(f"Duplicate node output {out}")
            seen_names.add(out)

    for n in graph.nodes:
        _check_op_specific(n, ctx)
        for inp in n.inputs:
            if inp and inp not in seen_names:
                ctx.errors.append(f"Dangling input {inp}")

    if ctx.errors:
        raise ValidationError(", ".join(ctx.errors))

    return True


async def check_model_async(model: Any, ctx: Optional[ValidationContext] = None):
    """Check a whole model for validity asynchronously."""
    return check_model(model, ctx)
